crates/ruff_server/src/session.rs

```diff
@@ -32,68 +33,102 @@ pub(crate) struct Session {
 pub(crate) struct DocumentSnapshot {
     resolved_client_capabilities: Arc<ResolvedClientCapabilities>,
     client_settings: settings::ResolvedClientSettings,
-    document_ref: workspace::DocumentRef,
+    document_ref: index::DocumentQuery,
     position_encoding: PositionEncoding,
-    url: Url,
 }
 
 impl Session {
     pub(crate) fn new(
         client_capabilities: &ClientCapabilities,
         position_encoding: PositionEncoding,
         global_settings: ClientSettings,
-        workspaces: Vec<(Url, ClientSettings)>,
-    ) -> crate::Result<Self> {
-        Ok(Self {
+        workspace_folders: Vec<(PathBuf, ClientSettings)>,
+    ) -> Self {
+        Self {
             position_encoding,
-            workspaces: workspace::Workspaces::new(workspaces, &global_settings)?,
+            index: index::Index::new(workspace_folders, &global_settings),
             global_settings,
             resolved_client_capabilities: Arc::new(ResolvedClientCapabilities::new(
                 client_capabilities,
             )),
-        })
+        }
     }
 
+    /// Creates a document snapshot with the URL referencing the document to snapshot.
     pub(crate) fn take_snapshot(&self, url: &Url) -> Option<DocumentSnapshot> {
+        let key = DocumentKey::from_url(url);
         Some(DocumentSnapshot {
             resolved_client_capabilities: self.resolved_client_capabilities.clone(),
-            client_settings: self.workspaces.client_settings(url, &self.global_settings),
-            document_ref: self.workspaces.snapshot(url)?,
+            client_settings: self.index.client_settings(&key, &self.global_settings),
+            document_ref: self.index.make_document_ref(key)?,
             position_encoding: self.position_encoding,
-            url: url.clone(),
         })
     }
 
-    pub(crate) fn open_document(&mut self, url: &Url, contents: String, version: DocumentVersion) {
-        self.workspaces.open(url, contents, version);
-    }
+    /// Updates a text document at the associated `key`.
+    ///
+    /// The document key must point to a text document.
+    pub(crate) fn update_text_document(
```

Comment:
What happens if it doesn't? Is this function subject to a panic? If so, I'd add it in the function docs.

Reply:
It throws an error - I'll add that to the docs.